flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/sources/DynamicSourceUtils.java

```diff
@@ -267,6 +302,20 @@ private static void validateScanSourceForStreaming(
 					scanSource.getClass().getName()
 				)
 			);
+		} else if (!changelogMode.containsOnly(RowKind.INSERT)) {
```

Comment:
The condition`!changelogMode.containsOnly(RowKind.INSERT)` does not indicate the source is a `CDC source which is non-upsert`

Reply:
This is in the else branch which means it doesn't match `!hasUpdateBefore && hasUpdateAfter` (upsert source).